### PR TITLE
Improve Task Runner Examples and Documentation

### PR DESCRIPTION
**Pull request by psfoley**
This PR:
- Makes changes to the `PyTorchTaskRunner` to wrap user `train_` and `validate_` functions. The impact of this is that for basic examples (including the quickstart), users don't need to understand or configure tensorkeys
- Adds on to the @kta-intel and @nammbash's quickstart and README content to show how to:
    1. Run a basic experiment for federated training
    2. Customize the deep learning model
    3. Customize training and validation tasks
    4. Configure the dataloader
    5. And rerun another experiment with those changes
- Several other small changes to the documentation

Items remaining before removing WIP tag:
- [x] Verifying all other PyTorch TaskRunner API examples (i.e. `torch_cnn_histology`, etc.) are compatible with base changes
- [x] Additional documentation updates

### CODE REVIEW

File: docs/get_started/quickstart.rst [hunk 1]
```diff
@@ -8,32 +8,359 @@ Quick Start
...
+   
+    ✔ OK
+   
+   Congratulations! You've run your first federation with OpenFL
+
+***************************
+Working with your own model
+***************************
+
+Now that you've run your first federation, let's see how to use replace the model used in the federation. After copying in the text above, you should be in the :code:`my_workspace` directory. Every workspace has a :code:`src` directory that contains the Task Runner, an OpenFL interface that defines the deep learning model, as well as the training and validation functions that will run on that model. In this case, the Task Runner is defined in :code:`src/taskrunner.py`. After opening it you'll see the following:
```
Comment (kminhta) [inline docs/get_started/quickstart.rst]: minor typo: "...let's see how **to use replace** the model used in the federation..."

File: docs/get_started/quickstart.rst [hunk 2]
```diff
@@ -8,32 +8,359 @@ Quick Start
...
-*********************************
-.. note::
+To begin we recommend installing OpenFL inside a python virtual environment. This can be done with the following:
 
-    Ensure you have installed the |productName| package.
+.. code-block:: console:
+    
+    pip install virtualenv
+    virtualenv ~/openfl-quickstart
+    source ~/openfl-quickstart
```
Comment (kminhta) [inline docs/get_started/quickstart.rst]: `source ~/openfl-quickstart/bin/activate`

**Review by kminhta — COMMENTED** [review]
Reviewed `quickstart.rst`. Great update to the quick start guide! Commands ran without issues. Very clear on how to get started and how to modify for custom use cases

File: docs/get_started/quickstart.rst [hunk 3]
```diff
@@ -8,32 +8,359 @@ Quick Start
...
-Hello Federation
-*********************************
-.. note::
+To begin we recommend installing OpenFL inside a python virtual environment. This can be done with the following:
 
-    Ensure you have installed the |productName| package.
+.. code-block:: console:
+    
+    pip install virtualenv
+    virtualenv ~/openfl-quickstart
```
Comment (nammbash) [inline docs/get_started/quickstart.rst:26]: I almost always suggest specific python versions that we are sure/tested against.
```virtualenv -p /usr/bin/python3.10 ~/openfl-quickstart/bin/activate```

File: docs/get_started/quickstart.rst [hunk 4]
```diff
@@ -8,32 +8,359 @@ Quick Start
...
+1. Setup Federation Workspace & Certificate Authority (CA) for Secure Communication
+2. Setup Aggregator & Initialize Federation Plan + Model
+3. Setup Collaborator 1
+4. Setup Collaborator 2
+5. Run the Federation
 
-Run the script
+.. code-block:: console
+
+   ############################################################################################
```
Comment (nammbash) [inline docs/get_started/quickstart.rst:41]: Suggestion: What do you two think on making these headings show up also in the terminal using echo?
``` 
echo "############################################################################################"
echo "# Step 1: Setup Federation Workspace & Certificate Authority (CA) for Secure Communication #"
echo "############################################################################################"

# Generate an OpenFL Workspace. This example will train a pytorch
# CNN model on the MNIST dataset
fx workspace create --template torch_cnn_mnist --prefix my_workspace
cd my_workspace

# This will create a certificate authority (CA), so the participants communicate over a secure TLS Channel
fx workspace certify

echo "#################################################################"
echo "# Step 2: Setup Aggregator & Initialize Federation Plan + Model #"
echo "#################################################################"

# Generate a Certificate Signing Request (CSR) for the Aggregator
fx aggregator generate-cert-request

# The CA signs the aggregator's request, which is now available in the workspace
fx aggregator certify --silent

# Initialize FL Plan and Model Weights for the Federation
fx plan initialize

echo "################################"
echo "# Step 3: Setup Collaborator 1 #"
echo "################################"

# Create a collaborator named "collaborator1" that will use data path "1"
fx collaborator create -n collaborator1 -d 1

# Generate a CSR for collaborator1
fx collaborator generate-cert-request -n collaborator1

# The CA signs collaborator1's certificate
fx collaborator certify -n collaborator1 --silent

echo "################################"
echo "# Step 4: Setup Collaborator 2 #"
echo "################################"

# Create a collaborator named "collaborator2" that will use data path "2"
fx collaborator create -n collaborator2 -d 2

# Generate a CSR for collaborator2
fx collaborator generate-cert-request -n collaborator2

# The CA signs collaborator2's certificate
fx collaborator certify -n collaborator2 --silent

echo "##############################"
echo "# Step 5. Run the Federation #"
echo "##############################"

# Run the Aggregator
fx aggregator start &

# Run Collaborator 1
fx collaborator start -n collaborator1 &

# Run Collaborator 2
fx collaborator start -n collaborator2

echo "Congratulations! You've run your first federation with OpenFL"
```

File: docs/get_started/quickstart.rst [hunk 5]
```diff
@@ -8,32 +8,359 @@ Quick Start
...
+    
+    pip install virtualenv
+    virtualenv ~/openfl-quickstart
+    source ~/openfl-quickstart
+    pip install openfl
 
-    See :ref:`install_package` for details.
 
-We will use the `"Hello Federation" python script <https://github.com/intel/openfl/blob/develop/tests/github/test_hello_federation.py>`_ to quickly create a federation (an aggregator node and two collaborator nodes) to test the project pipeline.
+Now you're ready to run your first federation! Copying these commands to your terminal will run a simple federation with two collaborators all on your local machine. These commands can be broken down into five steps, which you can read more about `here <../about/features_index/taskrunner.html#step-1-create-a-workspace>`_
```
Comment (nammbash) [inline docs/get_started/quickstart.rst]:
```suggestion
Now you're ready to run your first federation! Copying these commands to your terminal will run a simple federation with an aggregator & two collaborators all on your local machine. These commands can be broken down into five steps, which you can read more about `here <../about/features_index/taskrunner.html#step-1-create-a-workspace>`_

```

**Review by nammbash — COMMENTED** [review]
Patrick! very nice job in attempting to make the quick start guide easy to follow. 

Just minor suggestions and few nits from me.

I am confident now that the inertia of starting a federation will be reduced. (greatly with workflow API)

File: docs/get_started/quickstart.rst [hunk 6]
```diff
@@ -8,32 +8,359 @@ Quick Start
...
+            for data, target in validation_dataloader:
+                samples = target.shape[0]
+                total_samples += samples
+                data, target = data.to(self.device), target.to(self.device, dtype=torch.int64)
+                output = self(data)
+                # get the index of the max log-probability
+                pred = output.argmax(dim=1)
+                val_score += pred.eq(target).sum().cpu().numpy()
+        
+        accuracy = val_score / total_samples
```
Comment (nammbash) [inline docs/get_started/quickstart.rst:291]: Suggest adding return metric also as you make a mention specifically in the comment below
```suggestion
        accuracy = val_score / total_samples
        return Metric(name='accuracy', value=np.array(accuracy))
```

File: docs/get_started/quickstart.rst [hunk 7]
```diff
@@ -8,32 +8,359 @@ Quick Start
...
+
+Each function is passed a dataloader, and returns a :code:`Metric` associated with that task. In this example the :code:`train_` function returns the Cross Entropy Loss for an epoch, and the :code:`validate_` function returns the accuracy. You'll see these metrics reported when running the collaborator locally, and the aggregator will report the average metrics coming from all collaborators. 
+
+*****************************
+Defining your own data loader
+*****************************
+
+Now let's look at the OpenFL :code:`PyTorchDataLoader` and see how by subclassing it we are able to split the MNIST dataset across collaborators for training. You'll find the follow defined in :code:`src/dataloader.py`.
```
Comment (nammbash) [inline docs/get_started/quickstart.rst]:
```suggestion
Now let's look at the OpenFL :code:`PyTorchDataLoader` and see how by subclassing it we are able to split the MNIST dataset across collaborators for training. You'll find the following defined in :code:`src/dataloader.py`.

```

**Review by nammbash — COMMENTED** [review]
few more suggestions

File: docs/get_started/quickstart.rst (hunk 6 above)
Comment (psfoley) [reply docs/get_started/quickstart.rst:291]: I missed this. Great catch!
